### PR TITLE
Updates based on most recent audit

### PR DESCRIPTION
**Pull request by DeFiFoFum**
**IAO**
- [x] Issue 1: "Governance Privileges" - We acknowledge that this is a risk for users, but we feel that we can protect users funds if there is a token or pricing issue during the IAO that cannot be reversed.  
- [x] Issue 2: "Reentrancy Guard" - Added for relevant functions
- [x] Issue 3: "Lack of usage of upgradeable dependencies" - Added upgradeable deps
- [x] Issue 4: "adminAddress is private" - Currently the `adminAddress` in `IAO.sol` and `IAOLinearVesting.sol` are public
- [x] Issue 5: "deposited amount could still be zero"- Added a check for 0 value transfers
- [x] Issue 6: "Unnecessary rounding" - Now doing math in line to avoid rounding issues
- [x] Issue 7: "Function Visibility" - Changed relevant functions to `external`
- [x] Issue 8: "Lack of events" - Added relevant events
- [x] Issue 9: "checks-effects-interactions" - Fixed functions which do not follow this pattern
- [x] Issue 10: "Limiting Gas usage" - Added a notice at the top of the contract to evaluate this value on new networks
- [x] Issue 11: "Typographical Errors" - Fixed all typographical errors except `offeringTokensVested` ONLY in `IAO.sol` for backward compatibility

**IAOLinearVesting**
- [x] Issue 1: "checks-effects-interactions" - Fixed functions which do not follow this pattern
- [x] Issue 2: "Lack of validation" - Added validation

### CODE REVIEW

File: contracts/IAOLinearVesting.sol [hunk 1]
```diff
@@ -196,7 +218,8 @@ contract IAOLinearVesting is ReentrancyGuard, Initializable {
     /// @notice Calculate a users allocation based on the total amount deposited. This is done
     ///  by first scaling the deposited amount and dividing by the total amount.
     /// @param _user Address of the user allocation to look up
-    function getUserAllocation(address _user) public view returns (uint256) {
+    /// @notice This function has been deprecated, but leaving in the contract for backwards compatibility.
+    function getUserAllocation(address _user) external view returns (uint256) {
```
Comment (ape-swap) [inline contracts/IAOLinearVesting.sol:222]: Is the expectation for the frontend to do the Math? 

What is the proposed alternative?

File: contracts/IAOLinearVesting.sol [hunk 2]
```diff
@@ -274,7 +296,7 @@ contract IAOLinearVesting is ReentrancyGuard, Initializable {
             uint256 offeringTokenTotalHarvest, 
             uint256 offeringTokenInitialHarvest,
             uint256 offeringTokenVestedHarvest, 
-            uint256 offeringTokensVested
+            uint256 offeringTokensVesting
```
Comment (ape-swap) [inline contracts/IAOLinearVesting.sol:299]: This renaming may affect frontend integrations.

Comment (apewukong) [reply contracts/IAOLinearVesting.sol:299]: Yeah, so this PR's coming https://github.com/ApeSwapFinance/apeswap-frontend/pull/779. it's test ready

File: contracts/IAOLinearVesting.sol (hunk 1 above)
Comment (apewukong) [reply contracts/IAOLinearVesting.sol:222]: I believe getUserAllocation function is not deprecated here. It's just hinted for the external usage, if I'm not mistaken.

Comment (DeFiFoFum) [reply contracts/IAOLinearVesting.sol:222]: Maybe `deprecated` is the wrong term, but it's deprecated in the sense that the contract no longer relies on its math.

File: contracts/IAOLinearVesting.sol (hunk 2 above)
Comment (ape-swap) [reply contracts/IAOLinearVesting.sol:299]: Nice!

Once step ahead @apewukong

**Pull request merged**